### DISCUSSION TITLE
Autoscaling Pluto server by customizing how Pluto workers are launched?

### DISCUSSION THREAD

**Discussion by ericphanson** [fonsp/Pluto.jl#1244] [Pluto development]
I hope this is the right place to ask this. I was wondering about this line:

https://github.com/fonsp/Pluto.jl/blob/aef57ba471f6250a434ca389c1895952dfb84fd8/src/evaluation/WorkspaceManager.jl#L142

and in particular if there would be a way to allow passing a [cluster manager](https://docs.julialang.org/en/v1/stdlib/Distributed/#Distributed.addprocs). My idea is to use [K8sClusterManager](https://github.com/beacon-biosignals/K8sClusterManagers.jl) to have a Pluto server whose workers each live in kubernetes pods. Practically, this would just mean sending
```julia
K8sClusterManager(1; settings...)
```
instead of `1` in `addprocs` (where `settings` could customize the CPU allotment, memory allotment, etc, as [shown in the docs](https://beacon-biosignals.github.io/K8sClusterManagers.jl/dev/examples/#Launching-workers)). This would allow an autoscaling Pluto server which can spin up more pods as users launch more notebooks which seems pretty cool.

I guess I should just fork Pluto and try it out locally to see if it works first (K8sClusterManager works with minikube for local testing), but I wanted to also discuss it somewhere to see what other people's thoughts where.

**Comment by ericphanson** [comment]
Hm, one thing I didn't think of is where the notebooks get saved. Having the server be remote from the workers might need more extensive changes than just the `addprocs`...

**Comment by fonsp** [comment]
Awesome idea! If https://github.com/beacon-biosignals/K8sClusterManagers.jl is a drop-in replacement for Distributed, then that would work straight away! This might also be useful for other applications.

Maybe we can add a 'hook' that other code/packages can hook into to change the `1` into something else?

**Reply by ericphanson** [reply]
Yeah, so Distributed has this ClusterManager interface (https://docs.julialang.org/en/v1/stdlib/Distributed/#Cluster-Manager-Interface), and `K8sClusterManager` conforms to that interface. So I think maybe one thing Pluto could do is have a `manager` be configurable and default to `Distributed.LocalManager`. Then that would be the exact same as now, but if you change the `manager` you could launch workers elsewhere. E.g. https://github.com/JuliaParallel/ClusterManagers.jl provides interfaces to a bunch of other cluster types as well.
